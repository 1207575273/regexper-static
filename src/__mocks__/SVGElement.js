import React from 'react';

import Base from 'components/SVG/Base';

class SVGElement extends Base {
  reflow() {
    return this.setBBox(this.props.bbox);
  }

  render() {
    return <text>Mock content</text>;
  }
}

export default SVGElement;
