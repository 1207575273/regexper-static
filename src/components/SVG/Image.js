import React from 'react';
import PropTypes from 'prop-types';

import style from './style';

import reflowable from './reflowable';

const namespaceProps = {
  'xmlns': 'http://www.w3.org/2000/svg',
  'xmlns:cc': 'http://creativecommons.org/ns#',
  'xmlns:rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#'
};
const metadata = `<rdf:rdf>
 <cc:license rdf:about="http://creativecommons.org/licenses/by/3.0/">
    <cc:permits rdf:resource="http://creativecommons.org/ns#Reproduction"></cc:permits>
    <cc:permits rdf:resource="http://creativecommons.org/ns#Distribution"></cc:permits>
    <cc:requires rdf:resource="http://creativecommons.org/ns#Notice"></cc:requires>
    <cc:requires rdf:resource="http://creativecommons.org/ns#Attribution"></cc:requires>
    <cc:permits rdf:resource="http://creativecommons.org/ns#DerivativeWorks"></cc:permits>
  </cc:license>
</rdf:rdf>`;

@reflowable
class Image extends React.PureComponent {
  static defaultProps = {
    padding: 10
  }

  state = {
    width: 0,
    height: 0
  }

  reflow() {
    const { padding } = this.props;
    const box = this.children[0].getBBox();

    return this.setStateAsync({
      width: Math.round(box.width + 2 * padding),
      height: Math.round(box.height + 2 * padding)
    });
  }

  containedRef = contained => this.children = [contained]

  svgRef = svg => this.svg = svg

  render() {
    const { width, height } = this.state;
    const { padding, children } = this.props;

    const svgProps = {
      width,
      height,
      viewBox: [0, 0, width, height].join(' '),
      style: style.image,
      ref: this.svgRef,
      ...namespaceProps
    };

    return <svg { ...svgProps }>
      <metadata dangerouslySetInnerHTML={{ __html: metadata }}></metadata>
      <g transform={ `translate(${ padding } ${ padding })` }>
        { React.cloneElement(React.Children.only(children), {
          ref: this.containedRef
        }) }
      </g>
    </svg>;
  }
}

Image.propTypes = {
  children: PropTypes.node,
  padding: PropTypes.number
};

export default Image;
