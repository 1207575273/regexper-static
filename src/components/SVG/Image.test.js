import React from 'react';
import { mount } from 'enzyme';

import Image from './Image';

import SVGElement from '__mocks__/SVGElement';

describe('Image', () => {
  test('rendering', async () => {
    const component = mount(
      <Image>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Image>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });
});
