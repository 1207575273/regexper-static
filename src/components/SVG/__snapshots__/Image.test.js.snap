// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image rendering 1`] = `
<Image
  padding={10}
>
  <svg
    height={120}
    style={
      Object {
        "backgroundColor": "#fff",
      }
    }
    viewBox="0 0 120 120"
    width={120}
    xmlns="http://www.w3.org/2000/svg"
    xmlns:cc="http://creativecommons.org/ns#"
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
  >
    <metadata
      dangerouslySetInnerHTML={
        Object {
          "__html": "<rdf:rdf>
 <cc:license rdf:about=\\"http://creativecommons.org/licenses/by/3.0/\\">
    <cc:permits rdf:resource=\\"http://creativecommons.org/ns#Reproduction\\"></cc:permits>
    <cc:permits rdf:resource=\\"http://creativecommons.org/ns#Distribution\\"></cc:permits>
    <cc:requires rdf:resource=\\"http://creativecommons.org/ns#Notice\\"></cc:requires>
    <cc:requires rdf:resource=\\"http://creativecommons.org/ns#Attribution\\"></cc:requires>
    <cc:permits rdf:resource=\\"http://creativecommons.org/ns#DerivativeWorks\\"></cc:permits>
  </cc:license>
</rdf:rdf>",
        }
      }
    />
    <g
      transform="translate(10 10)"
    >
      <SVGElement
        bbox={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      >
        <text>
          Mock content
        </text>
      </SVGElement>
    </g>
  </svg>
</Image>
`;
